augur/filter.py

```diff
@@ -1,383 +1,261 @@
-            group_by = set(['date' if cat in ['year','month'] else cat
-                            for cat in args.group_by])
-            missing_cats = [cat for cat in group_by if cat not in meta_columns]
-            if missing_cats:
-                print("WARNING:")
-                if any([cat != 'date' for cat in missing_cats]):
-                    print("\tSome of the specified group-by categories couldn't be found: ",
-                          ", ".join([str(cat) for cat in missing_cats if cat != 'date']))
-                if any([cat == 'date' for cat in missing_cats]):
-                    print("\tA 'date' column could not be found to group-by year or month.")
-                print("\tFiltering by group may behave differently than expected!\n")
-
-            if args.priority: # read priorities
-                priorities = read_priority_scores(args.priority)
-
-            # subsample each groups, either by taking the spg highest priority strains or
-            # sampling at random from the sequences in the group
-            seq_subsample = []
-            for group, sequences_in_group in seq_names_by_group.items():
-                if args.priority: #sort descending by priority
-                    seq_subsample.extend(sorted(sequences_in_group, key=lambda x:priorities[x], reverse=True)[:spg])
-                else:
-                    seq_subsample.extend(sequences_in_group if len(sequences_in_group)<=spg
-                                         else random.sample(sequences_in_group, spg))
-
-            num_excluded_subsamp = len(seq_keep) - len(seq_subsample)
-            seq_keep = seq_subsample
-
-    # force include sequences specified in file.
-    # Note that this might re-add previously excluded sequences
-    # Note that we are also not checking for existing meta data here
-    num_included_by_name = 0
-    if args.include and os.path.isfile(args.include):
-        with open(args.include, 'r') as ifile:
-            to_include = set(
-                [
-                    line.strip()
-                    for line in ifile
-                    if line[0]!=comment_char and len(line.strip()) > 0
-                ]
+    """
+    Apply exclusion, inclusion, and subsampling rules to produce a subset of input data.
+
+    This command supports exclusion or inclusion rules TODO TODO TODO
+
+    Also subsampling TODO
+    """
+    Filterer(args).run()
+
+
+class Filterer:
+    def __init__(self, args):
+        self.args = args
+        self.sequence_file = SequenceFile(args.sequences, args.metadata)
+
+        self.check_vcftools()
+
+        self.sequences = self.sequence_file.sequences
+
+    def run(self):
+        self.filter_sequences()
+        self.print_summary()
+        self.write_output_file()
+
+    def translate_args(self):
+        """
+        Translate legacy CLI arguments to the generic scheme.
+
+        This is transitional code while the legacy scheme is still supported.
+        """
```

Comment:
So, this whole concept, and its use elsewhere in the matchers and matcher selection, is what I'm not sold on for a few reasons -
 
1. First, just to be clear, as someone who's not a maintainer of the project, I couldn't give a shipit on this kind of change even if I loved it - this is pointing to a huge change in the `augur filter` interface, and I can't make that call.
2. Details of the implementation here are leading to things like string parsing of arguments in the 'matchers', which makes the matchers harder to write and understand and also is generally error prone.
3. I think the solution is over-optimizing for "drop a filter in and it just magically works" - I'm not convinced that's a scenario frequent enough to require this level of work. I also don't think the proposed solution is actually easier to understand or implement than just the naive method of argument parsing and manually adding filters where they belong.
4. Again, the initial goal here was to refactor filter.py to make it testable - I think the filters as classes model can work here, but this review is going beyond that right now.

I mentioned this in the previous review - the work that's been done to make the filters into standalone classes is good and improves testability and comprehensibility. I'm on board with that. I'm not on board with the new approach to arguments, and I think it'd be better to break the review into two phases.

Reply:
You may be right about overoptimization for generic filters. I fell in love with the idea of `filter.py` not having any knowledge of the filter classes' internals. Now with a few days of distance, that does seem like a lot of change for questionable benefit. Thanks for the reality check :grin:.

I'll do what you suggested earlier and restore the argument parsing to `Filterer`, which can just instantiate the filter objects directly.